Redesigning Design Methods with Bayesian Optimisation

m-bone opened the pull request:
Redesigning method.py to allow for key usability improvements namely:

- User should be able to change sampling/optimising method without changing the sim_creation function
- User should be able to run single simulations or batches of simulations without changing the sim_creation function

Included within this is the addition of Bayesian optimisation and the creation of a MethodOptimise subclass to enable rapid development of additional optimisation methods e.g. genetic algorithms.

This will require refactoring to reduce achieve this setup without significant compromising the current user experience. Principle change will be the requirement of pre_fn and post_fn functions for creating and processing simulated data.

m-bone committed:
Design optimization plugin update

Added Bayesian optimization, genetic algorithms, and particle swarm
optimization methods
Removed MethodRandom and MethodRandomCustom as had become redundant
Expanded DesignSpace with new support methods estimate_cost and
summarize
Redesigned interface with plugin to make it easier for users to batch
simulations for different design search methods
Automated batching of simulations when user provides pre and post
processing functions
Expanded testing for all new features

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -57,116 +70,236 @@ def get_num_points(fn_args: Dict[str, tuple]) -> int:
+            return batch.run()
+
+    @staticmethod
+    def _check_pre_output(self, parameters: Tuple[ParameterType, ...], pre_fn: Callable):
```
Might be cleaner to formulate this as 
```py
def pre_output_is_sim_like(parameters: Tuple[ParameterType, ...], pre_fn: Callable) -> bool:
    """Whether the ``pre_fn`` returns an arrangement of ``td.Simulation``."""
```
instead of relying on returning specific strings.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -57,116 +70,236 @@ def get_num_points(fn_args: Dict[str, tuple]) -> int:
+
+        # If a single simulation is supplied
+        if isinstance(sims, Simulation):
+            return web.run(sims, **run_kwargs)
```
there's an equivalent to `Batch` called `web.Job`, maybe we can use that for consistency.

```py
if isinstance(sims, Simulation):
    job = web.Job(simulation=sims, ...)
    return job.run()
else:
    batch = web.Batch(simulations=sims)
    return batch.run()
```

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -57,116 +70,236 @@ def get_num_points(fn_args: Dict[str, tuple]) -> int:
+        # Uses batches if multiple simulations have been supplied
+        else:
+            batch = web.Batch(
+                simulations=sims, simulation_type="tidy3d_design_testing", **run_kwargs
```
simulation_type probably should go into the `Simulation` case too. An easy way to do this is at the top of the function
```py
run_kwargs.update(simulation_type="tidy3d_design")
```

tylerflex reviewed (commented):
didnt get a ton of time to fully digest it today but we can discuss tomorrow. So far, it's looking good though!

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +105,117 @@ def run(self, function: Callable, **kwargs) -> Result:
+                sub_dict = {batch_out[task_name] for task_name in sub_structure}
+                data.append(fn_post(sub_dict))
+
+        elif all(isinstance(sim, List) for sim in pre_out.values()) and all(
```
might be easier if instead of copying the above part, if this condition is met, then we just convert `sim` to a dict of dicts and then either recurse or just continue to the dict of dict case?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +105,117 @@ def run(self, function: Callable, **kwargs) -> Result:
+        def fn_combined(args_list):
+            sim_dict = {idx: fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
+            data = self.fn_mid(sim_dict, fn_post)
+            # post_out = [fn_post(val[1]) for val in data.items()]
```
reason for removing this? it would seem to me that if `fn_mid` just returned something dictionary-like, we should be ok?

In the various cases handled:

| pre_out type   | fn_mid return type |
| -------- | ------- |
| dict of Sim  | BatchData    |
| dict of dict of Sim | dict of BatchData     |
| dict of list of Sim   | dict of BatchData (same)    |
| dict of anything else | dict of anything else |

and all of these return types can just be handled in `fn_combined` like

```py
pre_out_dict = {idx: fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
post_in_dict = self.fn_mid(pre_out_dict)
return [fn_post(val) for _, val in post_in_dict.items()]
```

Users just need to know that if their `fn_pre` returns a `list` or `dict` of `Simulation`, their `fn_post` should expect a `BatchData`. Otherwise I think this works?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +105,117 @@ def run(self, function: Callable, **kwargs) -> Result:
+    def fn_mid(pre_out: Any, fn_post: Callable) -> Any:
+        """A function of the output of ``fn_pre`` that gives the input to ``fn_post``."""
+
+        # Handle most common case where fn_combined builds a dict of sims
```
might be possible eventually to just handle this super generally. For example:

* we're given a nested dictionary of stuff. 
* We iterate through this nested dictionary:
    * any `Sim` we find, goes in a `flattened_sims` dict.
    * mark where it exists in the input.
* We run all of the simulations 
    * in a batch, if there are more than 1
    * in a web.run() or Job.run() if there are 1
    * do nothing otherwise.
* We parse the remaining BatchData
    * just return it if the original dictionary was not nested.
    * split the BatchData and insert them into the result dictionary.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +105,117 @@ def run(self, function: Callable, **kwargs) -> Result:
+        return fn_combined
+
+    @staticmethod
+    def fn_mid(pre_out: Any, fn_post: Callable) -> Any:
```
I think we should also consider the case where configures their `pre_fn` to return a `Batch` or `Job` already loaded. We can just run them directly.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +105,117 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        else:
+            # user just wants to split into pre and post, without tidy3d I guess
+            data = [fn_post(val[1]) for val in pre_out.items()]
```
if we handle things with the fn_post in the fn_combined, dont need this or the lines 210-213 above.

tylerflex reviewed (commented):
Hey @m-bone just went through. Looks like it's getting pretty close. A few things:

1. I think I still feel like it would make more sense to handle `fn_post` after `fn_mid`. In my mind, if we have `fn_mid` return something `dict-like` (ie either a `dict` or a `BatchData`) then we can simply iterate through it and feed values to `fn_post`, this would retain the memory saving of `BatchData.items()`, while I think simplifying the code. Let me know if you see any problem with that?
2. I think we still need to nail down some kind of well defined acceptable output type -> input type relationship for `fn_pre` and `fn_post`. The current code implicitly seems to handle

| fn_pre return type    | fn_post arg type | notes 
| -------- | ------- | -------- |
| Simulation  | SimulationData    | line 170
| dict[Any, Simulation] | dict[Any, SimulationData]     | line 191
| list[Simulation] | list[SimulationData]     | line 213
| Any | Any | line 217

I think we list need to consider a few things:
A: what if `fn_pre` returns a mixed `dict` or `list`? say some `Simulation` and some `float`? for example, they want to pass some computed value to their fn post? Do we allow this? how do we handle it more generally?
B. should be able to trivially handle the `fn_pre` returning `Batch` or `Job` directly I think.
C. I recall now that in the original design plugin, a dict output of `fn_pre` actually gets input as `kwargs` to `fn_post` and a `list` output gets passed as positional `args`. see after cell [11]. https://docs.flexcompute.com/projects/tidy3d/en/latest/notebooks/Design.html
I think we just need to decide how to fill out this table once and for all.. Here's my proposal

in a general sense

| fn_pre return type    | fn_post arg type | notes 
| -------- | ------- | -------- |
| list  | *args  | 
| dict  | **kwargs  | 
| other | other |

and then within individual data, basically
| fn_pre return type    | fn_post arg type | notes 
| -------- | ------- | -------- |
| Simulation  | SimulationData  | 
| Batch  | BatchData | 
| Job | SimulationData |
| anything else | itself | 

Then we can say that if there's some special cases we handle to speed things along..
* if a `list` or `dict` of stuff, any `Simulation`-like data gets combined into one batch run, but then passed to `fn_post` using the tables above.

That's pretty complex but I'll comment out some specifics to discuss

tylerflex commented:
| `fn_pre` return    | `fn_mid` does | `fn_post` call 
| -------- | ------- | -------- |
| `1.0`  | nothing | `fn_post(1.0)` |
| `[1,2,3]` | nothing | `fn_post(1,2,3)` | 
| `{'a': 2, 'b': 'hi'}` | nothing | `fn_post(a=2, b='hi')`|
| `Sim()`  | `web.run()` | `fn_post(SimData())` | 
| `Batch()`  | `web.Batch.run()` | `fn_post(BatchData())` | 
| `[Sim(), Sim()]`  | `web.Batch().run()` | `fn_post(SimData(), SimData())` | 
| `[Sim(), 1.0]`  | `web().run()` | `fn_post(SimData(), 1.0)` | 
| `[Sim(), Batch()]`  | `web.Batch().run()` | `fn_post(SimData(), BatchData())` | 
| `{'a': Sim(), 'b': Batch(), 'c': 2.0}`  | `web.Batch().run()` | `fn_post(a=SimData(), b=BatchData(), c=2.0)` | 

How does this sound?

m-bone commented:
Noticed another change for optimisers compared to samplers: the user may want to return the raw output data from a simulation run whilst the optimiser will only take a single float value. I think it would be sensible to allow the user to output the raw data as a list/dict and we combine it within the results df, whilst the optimiser just gets passed the float. The two output cases would be:

```
def td_post(sim_data):
    mnt1 = sim_data["field1"]
    mnt2 = sim_data["field2"]
    avg_field = (mnt1 + mnt2) / 2

    return avg_field

def td_post_raw_out(sim_data):
    mnt1 = sim_data["field1"]
    mnt2 = sim_data["field2"]
    avg_field = (mnt1 + mnt2) / 2

    return [avg_field, [mnt1, mnt2]]
```
We can include a MethodOptimise method that can deal with float or list output from the function and let the user know the requirement is that it must output a float or a list `list[0]` is the float the optimiser needs. We can then add the raw output to the results df, and where the user outputs a dict we can set the column labels as the keys.

Is this too much pressure on the user? I feel this output is something people will want and it fits well within our existing API design.

tylerflex commented:
Hey Matt, I think this is a good thing to consider. The way that it’s handled in jax and autograd is to support a `has_aux : bool = False` flag. When `has_aux is False`, the objective function is assumed to simply return a `float`. However, when `has_aux is True`, the objective function is assumed to return a `tuple` of length 2 where the first value is assumed to be the value being optimized (`float`), and the 2nd item can be `Any`thing the user wants. We could treat it like regular `MethodSample` output and somehow provide those `Results` at the end of the optimization run?

So yea what makes sense to me is to consider these two possible output possibilities to the `MethodOptimize`? (Where the aux part is handled like it would be in any other method). And I think we could consider putting this `has_aux` (or whatever we decide to call it) as a `pydantic.Field` inside of `MethodOptimize`  if you agree?

m-bone commented:
With regards to the table I've implemented tests that validate all cases except

| `fn_pre` return    | `fn_mid` does | `fn_post` call 
| -------- | ------- | -------- |
| `Sim()`  | `web.run()` | `fn_post(SimData())` | 

For Sim() I think the correct result is to run `web.Batch.run()` and handle as many sims as is appropriate for the sampler/optimiser. This is the auto-batching that we've baked into this new API.

I'm also not convinced on allowing users to run their own batches using pre-post function inputs. Fn_mid doesn't currently support this so would need to be expanded. Creating a Batch feels like an unnecessary extra step because the user can just give us the dict they used if they wanted us to handle the batch. If we flatten the batch to take advantage of running sims in parallel, we'd likely lose whatever the user has specified as batch parameters. If the user really does want to manage their own batches it can be done with a single function.

tylerflex commented:
Just to respond to this officially / for reference (we kind of already discussed earlier):

> Creating a Batch feels like an unnecessary extra step because the user can just give us the dict they used if they wanted us to handle the batch.

This is true except for the memory advantage of `BatchData`.

> If we flatten the batch to take advantage of running sims in parallel, we'd likely lose whatever the user has specified as batch parameters.

Yea this is true too.

>  If the user really does want to manage their own batches it can be done with a single function.

Yes, but if the single function is used, then we can't take advantage of the parallelism over the arg_list (eg if each individual in the genetic algorithm population runs a batch, we'd have to run each batch sequentially).

At some point, it might be worth building in some special cases to allow Batches of Batches.. with `BatchData` argument types to `fn_post`. But only under specific cases? we could also build this functionality directly into `Batch` itself 🤔

m-bone commented:
Hi @tylerflex, I think this is now ready for the first round of review. 

Sampler now ignores requirements for specific outputs and will take any object, whilst optimisers will now reliably throw an error for an unsupported outputs early on in the the execution.

Testing has been expanded and is now covering >95% of the code. I believe I've implemented all the things we've discussed, as well as some extras so fire away if there's any questions!

tylerflex commented:
Thanks @m-bone ! I'm flying home today but will review as much as I can and maybe will have a review by tomorrow (my) morning or early afternoon.

tylerflex commented on tests/test_plugins/test_design.py:
```diff
@@ -175,113 +298,313 @@ def scs(radius: float, num_spheres: int, tag: str) -> float:
     with pytest.raises(ValueError):
-        tdd.MethodRandomCustom(num_points=5, sampler=SamplerNoRandom())
+        no_random = tdd.MethodRandomCustom(num_points=5, sampler=SamplerNoRandom)
+        no_random.get_sampler(parameters)
```
pretty nit picky, but technically I think we should initialize `no_random` before we enter the `pytest.raises` context. This way we can ensure that the `get_sampler` is the bit raising the error, not the initializer.

m-bone replied:
Nit picky but sensible! I've changed all the other tests so they all follow this structure.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -45,39 +45,81 @@ class DesignSpace(Tidy3dBaseModel):
         fn_values: List[Any],
         fn_source: str,
         task_ids: Tuple[str] = None,
-        batch_data: BatchData = None,
```
what is the reasoning behind removing batch data? this was a requested feature (being able to access the batch associated with the design plugin, if pre/post were used). it's possible aux_data can help with some of this but just think we should discuss since it's an API change.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -45,39 +45,81 @@ class DesignSpace(Tidy3dBaseModel):
 
-        fn_args_coords = tuple(fn_args.values())
+        fn_args_coords = tuple(
+            [[arg_dict[key] for arg_dict in fn_args] for key in fn_args[0].keys()]
```
confused, so is `fn_args` no longer a `Dict[str, tuple]` like in the type annotation?

m-bone replied:
Yes sorry forgot to update the typing. I'd made the change as all the optimisers produce inputs to the fitness function in the style of `dict[str, Any]`, and it didn't really make a difference to the samplers.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
 
-        return str(fn_source_pre) + "\n\n" + str(fn_source_post)
-
-    def run_batch(
```
for backwards compatibility, do we want to just keep this but have it wrap the `run_pre_post` with a `log.warning()` about the new syntax?

m-bone replied:
Changed. Discussion for later, are there a writing recommendations or a standardised "tone" for writing in tidy3D?

tylerflex replied:
Hm, not really, best is to take inspiration from the existing warnings and things. but for this, the tone I'd take would be something like.

```
In version 2.8.0, the 'run_batch' method is replaced by 'run_pre_post'. While the original syntax will still be supported, future functionality will be added to the new method moving forward.
```

In general I think we want to avoid referencing ourselves or the user directly. so for example "We recommend you do this" -> "Is is recommended to do this".  (had to correct my own suggestion).

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+            task_ids=self._task_names,
         )
+
+    def run_single(self, fn: Callable):
```
add return type annotations

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        return fn_combined
+
+    def fn_mid(self, pre_out: Any) -> Any:
```
do we want to make this private? ie `_fn_mid`? or do we expect the user to use this?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        # Convert list of sim inputs to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, List) for sim in pre_out.values()) and any(
```
just normal `list` is fine / preferred. I think some of the typing stuff is being phased out in newer python versions.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        return fn_combined
+
+    def fn_mid(self, pre_out: Any) -> Any:
```
`pre_out` is a `dict[int, Any]` correct? and then this returns something `dict`-like? or more specifically, something that can be iterated over its `.items()`?

m-bone replied:
Yeah you're right. Should the return just be typed as `dict` or is there a clearer way because it can be a BatchData in the case of `dict[int, Simulation]` inputs?

tylerflex replied:
I think we can type it as `dict[int, Any]` or we can say `Union[dict[int, Any], BatchData]`

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+        # Convert list of sim inputs to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, List) for sim in pre_out.values()) and any(
+            isinstance(sim, Simulation) for sim in pre_out[0]
```
so the idea here is that if we are given any `list[Simulation]`, we convert them to `dict[int, Simulation]`, run them all together in a batch and then return the `list[SimulationData]`?

Right now this assumes that if we are given `pre_out` where all values are `list` and any of the first `list` contains simulations that everything is a simulation?

What is the intended behavior if 
(a) we have a mix of simulations and other types in the list? 
(b) some of the lists contain simulations, but other lists don't?

I think we need to define the behavior and then handle it here (even if it means erroring), just to catch all edge cases.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -45,39 +45,81 @@ class DesignSpace(Tidy3dBaseModel):
+            aux_values=aux_values,
         )
 
+    def _create_name_generator(self):
```
this name generator is a bit tricky to follow. I'm trying to understand whether there's a way to avoid using a generator here.

Would it be possible to define a 
```py
def _task_name(self, counter: int) -> str:
```
and then pass the counter?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        # Handle most common case where fn_combined builds a dict of sims
+        if all(isinstance(sim, Simulation) for sim in pre_out.values()):
+            named_sims = {next(self._name_generator): sim for sim in pre_out.values()}
```
related to the generator discussion, what if we just maintained the counter in this `fn_mid` and passed it to a `self._task_name(counter)` function?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -45,39 +45,81 @@ class DesignSpace(Tidy3dBaseModel):
+
+    _dict_name: str = pd.PrivateAttr(default=None)
+
+    _task_names: list = pd.PrivateAttr(default=[])
```
As a general philosophy, we avoid using mutable values within these tidy3d components. The reason for this is that it's generally much harder to reason about things if one has to keep track of state. 

I put some assorted thoughts about removing the generator and also the task_names list. it would probably save us a lot of headaches in the future based on my experience.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+        # Handle most common case where fn_combined builds a dict of sims
+        if all(isinstance(sim, Simulation) for sim in pre_out.values()):
+            named_sims = {next(self._name_generator): sim for sim in pre_out.values()}
+            self._task_names.extend(list(named_sims.keys()))
```
like the generator, I'd rather just have the `task_names` be a variable in the scope of `fn_mid` rather than a mutable attribute of the `DesignSpace`. Is this possible? we can have `fn_mid` return both `data, task_names` and then package this in the `Results` if this makes things easier.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        # Can "index" dict here because fn_combined uses idx as the key
+        elif all(isinstance(sim, Dict) for sim in pre_out.values()) and any(
+            isinstance(sim, Simulation) for sim in pre_out[0].values()
```
same comment here about only checking first value for sim-ness. we should make sure to handle all three edge cases 1. all sims, 2. no sims, 3. mixed. even if it means erroring for case 3.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+        ):
+            # Flatten dict of dicts whilst storing combined keys for later
+            flattened_sims = {}
+            original_structure = {}
```
a handy trick is to use
```py
from collections import defaultdict

...

original_structure = defaultdict(dict)
```

then you can just assume that the key exists and is an empty dict, eg.
```
original_structure[dict_idx][sub_dict_idx] = blah
```
without needing to deal with `sub_structure`

might help simplify things.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+            # Run sims with flattened dict
+            named_sims = {}
+            for sim, key_name in zip(flattened_sims.values(), dict_keys):
+                self._dict_name = key_name
```
if we get rid of the generator, we can also remove the need for this? and also the `dict_keys`?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,78 +144,146 @@ def run(self, function: Callable, **kwargs) -> Result:
+            translation_dict = dict(zip(named_sims, flattened_sims))
+            self._task_names.extend(list(named_sims.keys()))
+            batch_out = web.Batch(simulations=named_sims, folder_name=self.folder_name).run(
+                path_dir=self.path_dir
```
might want to consider adding the `simulation_type` here, which lets us track if this was run, eg `simulation_type="tidy3d_design"`. I think this was in the original implementation

tylerflex commented on tidy3d/plugins/design/parameter.py:
```diff
@@ -30,7 +30,14 @@ class Parameter(Tidy3dBaseModel, ABC):
+                model=Parameter,
+                errors=[
+                    pd.error_wrappers.ErrorWrapper(
+                        ValueError("Supplied 'values' were not unique."), "/Parameter"
```
what is this? haven't seen this ErrorWrapper before

tylerflex replied:
in fact, it should be fine to just `raise ValueError(("Supplied 'values' were not unique.")` here and below. I think we phased out using `pd.ValidationError` in validators.

m-bone replied:
Pd was failing all the existing `pd.ValidationError` calls because it expects these error wrappers. It's a bit unnecessary here for how we're using it so I'll replace them all with `ValueError` instead

tylerflex commented on tidy3d/plugins/design/result.py:
```diff
@@ -61,23 +62,18 @@ class Result(Tidy3dBaseModel):
     )
 
-    task_ids: Tuple[Tuple[str, ...], ...] = pd.Field(
+    task_ids: list[str] = pd.Field(
```
we use `tuple[str, ...]` because it's immutable.

m-bone replied:
That's fair, will address this once we settle on the name generator approach

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,150 +29,88 @@ class Method(Tidy3dBaseModel, ABC):
-                "Convert these to 'tuple' for a workaround."
-            )
+        for param in parameters:
+            if type(param) == tdd.ParameterInt:
```
maybe `isinstance(param, tdd.ParameterInt)` is better, in class we subclass later.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -1,17 +1,23 @@
+from pyswarms.single.global_best import GlobalBestPSO
 
-from ... import web
+import tidy3d.plugins.design as tdd
```
not sure about this, doesn't it lead to some circular import? since `tdd` imports from this file? surprised this works.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -180,23 +124,430 @@ def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, A
+    )
+
+    kappa: Optional[pd.PositiveFloat] = pd.Field(
+        title="Kappa parameter for the Gaussian processor.",
```
for optional, we should set a default value, even if it's just `None`

m-bone replied:
It does have a default, I've just included it as a keyword argument below.

tylerflex replied:
ah I see! that's maybe a better way to do things to be honest. but since we generally put the default value as first argument in the fields, maybe can you move this kwarg to be the first one so it's more clear internally?

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -180,23 +124,430 @@ def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, A
+        )
+
+        # Create log and update
+        logger = JSONLogger(path="./logs")
```
should we just use our logger from `tidy3d/log.py`? or is it incompatible?

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -180,23 +124,430 @@ def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, A
+            _store_parameters.append(ga_instance.population.copy())
+            _store_fitness.append(ga_instance.last_generation_fitness)
+            best_fitness = ga_instance.best_solution()[1]
+            print(
```
for consistency, let's use `rich` similar to the web API. see `tidy3d/web/api/webapi.py` for examples on using `rich.Console`.

m-bone replied:
Ah that's interesting. If I'm honest, I think there's later discussion needed to finalise what we will output to the user as there's a huge amount of metadata that can be retained for the optimisers that is currently ignored

tylerflex replied:
sounds good

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -180,23 +124,430 @@ def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, A
+            )
+
+        # Make param names available to the fitness function
+        global _param_keys
```
We should avoid this if at all possible.  Can we just pass these to our fitness function as kwargs?

m-bone replied:
Unfortunately not, neither PyGAD or PySwarms allow their fitness function definitions in include any other arguments except for the ones it expects. A lot of their examples show them using globals or just defining variables in the same namespace as the function.

tylerflex replied:
Hm, so it doesn't work if you define `param_keys` before the fitness function?
```
_param_keys = [param.name for param in parameters]

def fitness_function(...)
    # use _param_keys_
```

tylerflex reviewed (commented):
Thanks @m-bone . I added my initial comments. Overall the structure is looking pretty good! my comments are mainly details. Let's just discuss each of them and iterate a few times on it.

m-bone commented:
@tylerflex when you're ready I think this is good for your review. I've implemented all changes and added estimate_cost, logging and QoL features. The DesignSpace delete method will coming next but I want to review the existing design wrt. updating state and then go from there.

I've removed the pre/2.8 commits that were included (again), so this is all just my design update

tylerflex commented:
Thanks @m-bone , enjoy your trip!

tylerflex commented on tests/test_plugins/test_design.py:
```diff
@@ -1,24 +1,61 @@
+    return BatchDataEmulated(data_dict=data_dict, task_ids=task_ids, task_paths=task_paths)
+
+
+def emulated_estimate_cost(self, verbose=True):
```
better to have this be deterministic. switch with a kwarg and just add a `@pytest.mark.parametrize()` for it?

m-bone replied:
I've refactored the testing and made a mock job.delete so the tests don't talk to the live servers. Tests read better now, this code is deterministic using pytest.parameterize with monkeypatch, and the 36 tests run in < 30 sec instead of > 5 mins.

tylerflex commented on tests/test_plugins/test_design.py:
```diff
@@ -175,113 +344,330 @@ def scs(radius: float, num_spheres: int, tag: str) -> float:
+    # Placed at the end of the test so that module is loaded for other checks
+
+    with pytest.raises(ImportError):
+        sys.modules[method_module_convert[sweep_method.type]] = None
```
can we use `monkeypatch.setattr` for this instead?

m-bone replied:
Setattr doesn't seem to do it because I need the whole module to ImportError later on. I've changed `sys.modules` with `monkeypatch.setitem` which I might be the best compromise.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -41,31 +49,48 @@ class DesignSpace(Tidy3dBaseModel):
         ...,
         title="Search Type",
         description="Specifications for the procedure used to explore the parameter space.",
+        discriminator="type",  # Stops pydantic trying to validate every method whilst checking MethodType
```
please use `TYPE_TAG_STR` from `tidy3d/components/base.py` (in case we change the string later).

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -41,31 +49,48 @@ class DesignSpace(Tidy3dBaseModel):
+    task_name: str = pd.Field(
+        None,
+        title="Task Name",
+        description="Task name assigned to tasks along with a simulation counter. Only used when pre-post functions are supplied.",
```
could you be more specific about what the actual task names will be? for example `{task_name}_{counter}` or something?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -41,31 +49,48 @@ class DesignSpace(Tidy3dBaseModel):
+    path_dir: str = pd.Field(
+        ".",
+        title="Path Directory",
+        description="Directory for files to output to. Only used when pre-post functions are supplied.",
```
maybe "Directory where simulation data files will be saved to." to be more specific.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -41,31 +49,48 @@ class DesignSpace(Tidy3dBaseModel):
 
-        fn_args_coords = tuple(fn_args.values())
+        fn_args_coords = tuple(
+            [[arg_dict[key] for arg_dict in fn_args] for key in fn_args[0].keys()]
```
does this assume all of the `fn_args` have the same `keys()`? Would it make more sense to not re-iuse the keys and instead make it a `Dict[str, tuple[Any, ...]]`?

m-bone replied:
Let's talk about this one because I'm a little lost as to your point

tylerflex replied:
so before, the `fn_args` was a dict of tuples. Now it's a list of dicts. The slightly awkward thing about the new approach is that each dict has its own keys (so in principle they could be different), whereas before it was just one set of keys.

Maybe I feel like the old way made a bit more sense, but perhaps there's a reason for the new way?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -86,7 +112,7 @@ def get_fn_source(function: Callable) -> str:
 
-    def run(self, function: Callable, **kwargs) -> Result:
+    def run(self, fn: Callable, fn_post: Callable = None, verbose: bool = True) -> Result:
         """Run the design problem on a user defined function of the design parameters.
```
will want to add a lot of explanation here for when to supply fn_post, what happens if various types are passed between fn and `fn_post`, etc. also in the other user facing functions.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+                data, sim_names = fn_mid(sim_dict, self.sim_counter)
+                self.sim_names.extend(sim_names)
+                self.sim_counter += len(sim_names)
+                post_out = [fn_post(val[1]) for val in data.items()]
```
typically, we'd do `fn_post(val) for _, val in data.items()`

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+    def _get_evaluate_fn_pre_post(self, fn_pre: Callable, fn_post: Callable, fn_mid) -> list[Any]:
+        """Get function that tries to use batch processing on a set of arguments."""
+
+        class Pre_Post_Handler:
```
still don't 100% understand why this is needed. if there's a way to avoid it, that would be preferred. 

what's the difference between this vs. just writing `fn_combined` as a method and having the `counter` and `sim_names` as variables in that method?

tylerflex replied:
So is the idea that we want subsequent iterations /epochs of eg the GA to have new counters?

For example, first epoch might be `sim1, sim2, sim3 ...`
but then we want the next epoch to be named `sim4`,`sim5` etc?

If so, what about just adding an ability to add a suffix to the task name when calling `evaluate`? and passing epoch?

`sim1_epoch1`, ...

m-bone replied:
> what's the difference between this vs. just writing `fn_combined` as a method and having the `counter` and `sim_names` as variables in that method?

Names have to be made within `fn_mid`, which sits within `fn_combined`, as they are used in the web.run calls to give sensible names to the servers. 

`fn_combined` can't store any variables because it has to be capable of being called once for the samplers, and being called many times by the optimizers. The optimizers call `fn_combined` within their fitness functions which are very difficult to get anything into or out of. We need something that sits in the same scope as `fn_combined` to capture names because we want to have these names for the Results object (although currently we're not doing much with them in Results, but it seems sensible to keep everything together). I'd rather not handle this for each method as it will bloat the method code and make adding new methods harder.

As to what the names are it's very open to interpretation. Currently the handler is also storing a counter and that is built into the name, but there are many different ways we could name things. What's the default naming conventions? Can the user not supply a task_id?

tylerflex replied:
> Names have to be made within fn_mid, which sits within fn_combined, as they are used in the web.run calls to give sensible names to the servers.

> fn_combined can't store any variables because it has to be capable of being called once for the samplers, and being called many times by the optimizers. The optimizers call fn_combined within their fitness functions which are very difficult to get anything into or out of. We need something that sits in the same scope as fn_combined to capture names because we want to have these names for the Results object (although currently we're not doing much with them in Results, but it seems sensible to keep everything together). I'd rather not handle this for each method as it will bloat the method code and make adding new methods harder.

Could the suffixes for example be passed from the method as it calls `fn_combined`? of course, they would need to be passed through to `fn_mid` and stuff. Would this cause more complication?

I think the current naming convention is ok, but it does sort of lock us into a specific one whereas allowing each method to pass whatever suffix (or even a function generating the task_name) might be more flexible? I dont understand all of the logic though, so I defer to your judgement on it.

> Can the user not supply a task_id?

Not sure what you mean? like the `Job.task_id`?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        return evaluate
+
+    def _get_evaluate_fn_pre_post(self, fn_pre: Callable, fn_post: Callable, fn_mid) -> list[Any]:
```
this returns I guess a `Pre_Post_Handler`, not list any?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+
+        # Convert list input to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, list) for sim in pre_out.values()):
```
what if just one of the pre_out.values is a list of sims?

m-bone replied:
Good idea. I've checked now and it's not possible for one of `pre_out.values` to be a list and the other to be anything else. `pre_out.values` is the return from the user supplied fn_pre which doesn't change during the run. If the return for fn_pre were say `return list, string` it would error before reaching this check.

tylerflex replied:
> If the return for fn_pre were say return list, string it would error before reaching this check.

I'm a bit confused, wouldn't `list, string` be acceptable?

m-bone replied:
So played around with this again and now it fails with a proper raise. The `list, string` example would be a tuple output from fn_pre which is not an accepted output. 

If it were `[list, string]` this would fail as well as the second element from fn_pre needs to be a list or dict.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+            """Recursively search for search_type objects within a dictionary."""
+            current_key = previous_key
+            for key, value in search_dict.items():
+                if len(previous_key) == 0:
```
`if not len(previous_key):`

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+        _find_and_map(pre_out, Simulation, simulations, naming_keys)
+        _find_and_map(pre_out, Batch, batches, naming_keys)
+
+        # Exit fn_mid here if td computation is required
```
if no computation is required?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+        _find_and_map(pre_out, Batch, batches, naming_keys)
+
+        # Exit fn_mid here if td computation is required
+        if len(simulations) == 0 and len(batches) == 0:
```
`if not simulations and not batches`

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+        _find_and_map(pre_out, Batch, batches, naming_keys)
+
+        # Exit fn_mid here if td computation is required
+        if len(simulations) == 0 and len(batches) == 0:
```
not sure we really need this? seems there are some edge cases eg if `batches is {}` but there are sims and similarly if `simulations is {}` but there are batches?

Would it make sense to just let the logic in the following sections take care of it?

m-bone replied:
This is here not so much because the next section can't deal with empty batches or simulations, but because the return statement needs to be different. Having this saves converting back to the original format of pre_out and handling no named_sims data.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
+        This function has been superceded by `run`, please use `run` for batched simulations.
+        """
+        log.warning(
+            "In version 2.8.0, the 'run_batch' method is replaced by 'run'."
```
could you mention that the fn_pre becomes `fn` and `fn_post` stays the same?

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
-            Optional directory in which to store the batch results.
+        if len(batch_kwargs) > 0:
+            log.warning(
+                "Batch_kwargs supplied here will no longer be used for within the simulation."
```
`'batch_kwargs'` (variable names in single quotes)

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
+
+    def _get_run_count(self) -> int:
+        # For MonteCarlo, Random, RandomCustom
+        if hasattr(self.method, "num_points"):
```
let's just use `isinstance(self.method, (MonteCarlo, Random, RandomCustom))`

generally want to avoid hasattr

tylerflex replied:
alternatively, we could provide a `def get_run_count(parameters: None) -> int` method to `Method` and then just implement it for each class and call it here. Would be probably the cleanest way to handle it.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
+        # Get output fn_pre for paramters at the lowest span / default
+        arg_dict = {}
+        for param in self.parameters:
+            if hasattr(param, "span"):
```
similar comment, it would be better if the `parameters` had some method returning perhaps some "sample" value for use in this. Rather than having to query them from `Design`.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
-            task_ids=task_ids,
-            batch_data=batch_data,
+            estimate = job.estimate_cost()
+            job.delete()  # Deleted as only a test with initial parameters
```
there's gotta be a better way to do this.. there should probably be a `web.estimate_cost(sim)` that does something similar but just doesn't retain the task.

tylerflex replied:
maybe we should file a ticket to work on it eventually

m-bone replied:
Yeah it does feel like estimate_cost could have an arg to delete on completion

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
+
+            return estimate
+
+        if isinstance(pre_out, Simulation):
```
consider making this a bit recursive to simplify things?

```py
def estimate_cost(pre_out):

    if isinstance(pre_out, Simulation):
         job = Job(simulation=pre_out)
         time.sleep(2.0) # not sure, maybe this will give the server enough time to compute it?
         return  estimate_cost(job)
    if isinstance(pre_out, (Job, Batch)):
         cost = pre_out.esimate_cost()
         pre_out.delete()
         return cost
    if isinstance(pre_out, dict):
        return estimate_cost(tuple(pre_out.values()))
    if isinstance(pre_out, (list, tuple)):
        return sum(estimate_cost(x) for x in pre_out)
    return 0.0
```
         



```

m-bone replied:
I've written this like this because it follows the same None handling logic as `Batch.estimate_cost()`. 

There's no race condition on our servers, they seem to queue and time out cost estimates correctly so I don't think a sleep is necessary.

We've got to break the lists and dicts down somewhere into component simulations and batches to take advantage of existing `Job.estimate_cost` or `Batch.estimate_cost` methods.

We can't return 0 for failed estimates as wouldn't people then believe the simulation is free?

tylerflex replied:
ok, I see. maybe this doesn't properly handle the None condition.

Btw the return 0 was meant to handle the case where pre_out was not sim-like.

If what you have before works, we can keep that.

m-bone replied:
Ah sorry, that makes more sense

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
+        else:
+            return round(per_run_estimate * run_count, 3)
+
+    def summary(self, fn_pre: Callable = None):
```
`-> None:`

I think if we go with the console.log approach, then this should be called `summarize`, or `print_summary`

it might be nice to get the summary as a `pd.BaseModel`, `dataclass`, or  `dict` as well

m-bone replied:
Changed to output the summary as dict so that the user has access to a the data as a variable. Works well for now, can be expanded in the future.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
+            if any(isinstance(param, ParameterInt) for param in self.parameters):
+                if any(isinstance(param, ParameterAny) for param in self.parameters):
+                    notes.append(
+                        "Discrete ParameterAny values are automatically converted to int values to be optimized.\n"
```
`'ParameterAny'`

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +322,175 @@ def run_batch(
+                    )
+
+                notes.append(
+                    "Discrete int values are automatically rounded as optimizers generate float predictions.\n"
```
`'int'`, `'float'`

tylerflex commented on tidy3d/plugins/design/result.py:
```diff
@@ -61,23 +62,24 @@ class Result(Tidy3dBaseModel):
-        "the parameter sweep function is split into pre and post processing and run with "
-        "'Design.run_batch()', otherwise is ``None``.",
+        title="Auxiliary values output from the user function",
+        description="The auxiliary return values from the design problem function. These weren't used to inform the optimizer, if one was used.",
```
more description about what these correspond to.

tylerflex commented on tidy3d/plugins/design/result.py:
```diff
@@ -61,23 +62,24 @@ class Result(Tidy3dBaseModel):
         "the parameter sweep function is split into pre and post processing and run with "
-        "'Design.run_batch()', otherwise is ``None``. "
-        "To access all of the data, see ``batch_data``.",
+        "'Design.run_batch()', otherwise is ``None``. ",
```
this syntax changed, so maybe "`Design.run()` with `fn_post` defined"? not sure.

tylerflex commented on tidy3d/plugins/design/result.py:
```diff
@@ -61,23 +62,24 @@ class Result(Tidy3dBaseModel):
+    opt_output: Any = pd.Field(
+        None,
+        title="Plots produced through design method",
+        description="Plot objects captured over the course of the design method run.",
```
using matplotlib? how do I plot these? some more description could be useful here. Also, I'm not sure if this will be serializable / uploadable (for future)

tylerflex commented on tidy3d/plugins/design/result.py:
```diff
@@ -186,8 +188,27 @@ def to_dataframe(self) -> pandas.DataFrame:
         columns = list(self.dims) + val_keys
 
+        if include_aux:
+            if self.aux_values is not None:
```
```py
if include_aux and self.aux_values:
```
or consider just warning if `include_aux` but not `self.aux_values`? not sure the intended API.

tylerflex commented on tidy3d/plugins/design/result.py:
```diff
@@ -186,8 +188,27 @@ def to_dataframe(self) -> pandas.DataFrame:
+        if include_aux:
+            if self.aux_values is not None:
+                # Can use [0] for aux keys as the function is assumed producing the same structure of output each run
+                if all(isinstance(auxs, dict) for auxs in self.aux_values):
```
sort of assumes that all of the `aux_datas[I]` are the same type and if dicts, they have the same keys?

m-bone replied:
This is true, though I think instances where this is false are very niche. If people have used `Result.combine` then it's been checked that the source and the columns have passed as the same. Particularly if the source is the same then you'd expect the aux to be identical.

Thinking about it the only way I can think of that breaks this is if people supply `fn_post` with different return statements which could be seen as valid for e.g. pseudocode

```
if sim_data["something"] == None:
    return fit_val, None
else:
    aux_dict = make_aux_dict(sim_data)
    return fit_val, aux_dict
```
Not really sure how to handle this case. It's probably unlikely so I suggest we wait and see if anyone does it. I think the current code could be modified easily enough if people wanted this functionality.

tylerflex replied:
> Thinking about it the only way I can think of that breaks this is if people supply fn_post with different return statements which could be seen as valid for e.g. pseudocode

yea I think this is what I had in mind as possibly breaking this code. I think if you dont want to handle it, we just need to just add a quick check that if aux_dict is supplied, that they all are the same type, and if that type is a dict it has the same keys, lists / tuples have the same length, etc. ? otherwise either raise an error that we dont know how to build the result using the aux_data provided or just warn and skip the aux data altogether?

m-bone replied:
I've got a light catch now where  `_extract_output` checks that all outputs have the same `type`.  This will catch crude multi-returns with an error that says raise a github issue if this is effecting you.

It is possible to sneak through an invalid output if the say the user has a return of `[float, list]` and `[float, dict]` but this will only fail when trying to create a dataframe. We're getting into the weeds with edge cases at this point, and it would be an easy enough patch to include this functionality if required.

tylerflex commented on tidy3d/plugins/design/result.py:
```diff
@@ -330,6 +351,21 @@ def delete(self, fn_args: Dict[str, float]) -> Result:
         new_coords.pop(index)
         new_values.pop(index)
 
+        # ParticleSwarm optimizer doesn't work with updated_copy
```
curious why not? what's the error?

m-bone replied:
If I remember rightly it's because parts of the PSO object are pickled which updated_copy fails to handle.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,150 +21,87 @@ class Method(Tidy3dBaseModel, ABC):
-    @abstractmethod
-    def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, Any]:
-        """Defines how the design parameters are sampled."""
+    def _force_int(self, next_point: dict, parameters: list):
```
return type `-> int`?

m-bone replied:
In a way yes but it's updating a dict in-place so I didn't think it warranted a return. What's the standard for anything in-place?

tylerflex replied:
oh I see. it would be `-> None:` then.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,150 +21,87 @@ class Method(Tidy3dBaseModel, ABC):
+                aux_out = []
+                for val in output:
+                    float_out.append(val[0])
+                    aux_out.append(val[1])
```
should this error if not all `len(val) == 2`?

m-bone replied:
Good point, it's now raising something here.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,150 +21,87 @@ class Method(Tidy3dBaseModel, ABC):
-            fn_output = fn(**fn_kwargs)
-            result.append(fn_output)
+    @staticmethod
+    def _flatten_and_append(list_of_lists: list[list], append_target: list):
```
return type, also docstring (for all methods).

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,150 +21,87 @@ class Method(Tidy3dBaseModel, ABC):
 
-        return fn_args, result, task_ids, batch_data
+        # None is aux_output and opt_output
+        return fn_args, results, None, None
```
what happened to the task_ids and batch_data?

m-bone replied:
I think this Github delete has picked up this modification from the original implementation of the method. The "None"s now refer to different things.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -175,28 +110,466 @@ class MethodGrid(MethodIndependent):
+    """A standard method for performing bayesian optimization search"""
+
+    initial_iter: pd.PositiveInt = pd.Field(
+        ...,
```
is there a good default for this?

m-bone replied:
Not especially, it's quite problem specific. I could set it to something vaguely sensible like 10, but I'd rather the user set it every time. If only because it would be confusing to the user if they didn't realise it had a default.

tylerflex commented on tidy3d/plugins/design/method.py:
no period here (or for Kappa title)

tylerflex commented on tidy3d/plugins/design/method.py:
it's confusing but for these (which go in our docs) we use double backslash quotes for variables.
Single quotes are for things that go to terminal output.

tylerflex replied:
can you double check all of the other instances?

m-bone replied:
Ah ok so it should be `ei` for example?

tylerflex replied:
not quite, 
```
``ei``
```

tylerflex commented on tidy3d/plugins/design/method.py:
single quotes on 'bayes_opt'.  also perhaps suggest they `pip install bayes_opt` or whatever the command would be .

tylerflex commented on tidy3d/plugins/design/method.py:
No periods needed in titles. (fix all)

tylerflex commented on tidy3d/plugins/design/method.py:
I'd probably prefer we just define a stop_critera_type as a literal "reach" or "saturate" and then the stop_critera number as an int / separate field

tylerflex replied:
reason is I'm not sure how this regex will work with our docs, it seems a bit tricky.

m-bone replied:
Fair point, changed to two fields and added a check that the fields combine successfully.

tylerflex commented on tidy3d/plugins/design/method.py:
consider linking to pyGAD docs here and in other instances.

m-bone replied:
Yeah I'd left it out here because the pygad docs are a total mess, everything important is essentially on one giant landing page **https://pygad.readthedocs.io/en/latest/pygad.html**. I guess it's more convenient than having the user Google it though.

tylerflex commented on tidy3d/plugins/design/method.py:
no need to make this `Optional` unless you want `None` to be allowed?

tylerflex commented on tidy3d/plugins/design/method.py:
docstring

tylerflex commented on tidy3d/plugins/design/method.py:
docstring

tylerflex replied:
one thing to consider is allowing  `**kwargs` to be passed here from `design.run()`. then the user could theoretically have these passed directly to `pygad.GA`. (similar for BO and other methods).

m-bone replied:
I think we leave it for the moment and see how users interact with it. If we allow kwargs through then people could accidentally overwrite settings assigned through pydantic fields, which feels like chaos. I'd rather go overboard and implement all the GA settings as pd fields.

tylerflex commented on tidy3d/plugins/design/method.py:
consider making this a `@property`. then the user could even call eg. `design_space.method.optimizer` and see this.

m-bone replied:
I'm slightly lost as to what this would allow? Why would the user need to interact with this?

tylerflex replied:
just wondering if the user would want to inspect the GA [class](https://pygad.readthedocs.io/en/latest/pygad.html#pygad-ga-class) before or after running. For example to see how things are set up or plot results.

For example

```py
des = DesignSpace(..., method=method_ga)

print(des.method.optimizer)

des.run()

```

m-bone replied:
Ah I see what you mean. They do have access to the optimizer object after the run as it's now output as Results.optimizer so people can plot and analyze it. 

I don't think we should give access to it before the optimization as it doesn't exist before method.run is called, and all the relevant parameters should be set with Fields when the method is created.

tylerflex replied:
ah ok that makes sense. 👍

tylerflex commented on tidy3d/plugins/design/method.py:
might want this to also happen only if verbose is True in the design.. if it's possible.

m-bone replied:
The way it's setup currently console can only not be None when Verbose is True so I figured it wasn't worth passing verbose through to method when it can just check the state of console.

tylerflex replied:
oh interesting. that works

tylerflex commented on tidy3d/plugins/design/method.py:
same comments about Optional. my understanding is that optional means something can be None. not necessarily that it has a default. Or to put another way, if something has a default, it doesn't need to be Optional, unless None is also an allowed value.

m-bone replied:
Thanks I hadn't realised this. Stripped out all the Optionals now.

tylerflex commented on tidy3d/plugins/design/method.py:
need to check if `self.rng_seed` is not none?

m-bone replied:
No numpy just ignores a 'None' and picks it's own seed instead.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -302,7 +681,13 @@ class MethodRandomCustom(AbstractMethodRandom):
 
-    @pd.validator("sampler")
+    sampler_kwargs: Any = pd.Field(
+        {},
```
hm, so dicts are mutable so this seems a bit unusual to me. What's the idea here? Also not sure about the hasattr business below.

Ideal: let's just remove the `MethodRandom`? or just have it be a simple class that inherits from `MethodMonteCarlo` with a warning (through a `pd.root_validator` if it's initialized?

m-bone replied:
Yeah we could just remove MethodRandom as the user can reimplement it in MethodRandomCustom come to think of it.

For the code above it's a dict because we needed to initialise the sampler in 'get_sampler' (so that random seed is set properly) and needed to pass the kwargs for the sampler through to this stage. Is there another way of doing this?

tylerflex replied:
I suppose this works. it just begs the question why not have it this way for eg the GA code? (pygad kwargs)? I think the idea is that having things listed out as fields makes things better documented, easier to read and write to file, and we can catch things at the validation stage before they cause errors.

Are the sampler kwargs the same over all of the possible random custom samplers?  would it make sense to just remove Custom Random samplers too and only support the builtin Monte Carlo ones?

tylerflex reviewed (commented):
Hey @m-bone , thanks for this. Looking pretty good! I had a lot of suggestions but I think it's getting there. 

My main concern at a high level is that I have trouble following every detail as things have gotten pretty complicated. I think a lot of this can't be avoided because we're trying to do a lot of things with this package. But in general if we can be really clear and explicit in the docstrings, descriptions, and errors, that will go a long way. This especially pertains to when there are edge cases that we handle (eg behavior A if given type a, behavior B when given type b). We should be super clear otherwise I worry it will be confusing to users. But overall I think it is pretty good. Thanks!

tylerflex commented on tests/test_plugins/test_design.py:
tests look good! Just ran coverage, seems the only thing we should probably add is some testing for when the fn_pre return type is a Batch as it seems not to be covered

m-bone replied:
Ahh that's only in estimate_cost for the batch. I've tweaked one of the simulation tests now so that it will hit that bit of code

tylerflex commented on pyproject.toml:
```diff
@@ -73,6 +73,11 @@ jax = { version = "0.4.25", extras = [
   "cpu",
 ], source = "jaxsource", optional = true }
 
+# design
```
I didnt have these installed and the error was nicely handled (only errored when the appropriate method was called)

tylerflex commented on tidy3d/plugins/design/parameter.py:
```diff
@@ -51,22 +51,26 @@ def _sample_grid(self) -> List[Any]:
         ...,
         title="Span",
-        description="(min, max) inclusive range within which the variable should be swept.",
+        description="(min, max) inclusive range within which are allowed values for the variable.",
```
should probably mention whether these bounds are inclusive? seems I forgot in the original description

tylerflex replied:
same comments below

tylerflex commented on tidy3d/plugins/design/parameter.py:
```diff
@@ -78,6 +82,9 @@ def span_size(self):
         return span_max - span_min
 
+    def sample_first(self) -> tuple:
+        return self.span[0]
```
please add docstring and do we want this to be private?

tylerflex replied:
same comments below

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
     @abstractmethod
-    def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, Any]:
-        """Defines how the design parameters are sampled."""
+    def get_run_count(self, parameters: list = None) -> int:
```
might make sense for this to be private and have some way to call it from the `DesignSpace` passing parameters, similar to `DesignSpace.run()` and `Method._run()`?

m-bone replied:
I've made it private but I feel like including it in DesignSpace like that just short circuits `DesignSpace.summarize()`, which is calling it and printing the result.

tylerflex replied:
yea potentially. if nothing else it might just help separate some of the logic out, but I'll leave it up to you

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
-                "Parameter sweep tool won't work with sets of lists, dicts or numpy arrays. "
-                "Convert these to 'tuple' for a workaround."
+                "Unrecognized output from supplied post function. The type of output varies across the output."
+                "Use of multiple return functions in fn_post is discouraged."
```
double check any instances of python code / variable names in the exceptions and put single quotes `'` (`'fn_post'`) here

m-bone replied:
What about pydantic field names? Is it ``stop_criteria_type`` or 'stop_criteria_type'?

tylerflex replied:
anything code related that's getting printed to terminal including field names should be in '. If it's going in docs (eg a docstring) it should be inside double `

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
-        fn_args, num_points = self._assemble_args(parameters)
+        else:
+            raise ValueError(
+                "Unrecognized output from supplied post function. Output should be a float or an iterable object."
```
same formatting as above for `float`, `int`, etc.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
-    ) -> BatchData:
-        """Create a batch of simulations and run it. Mainly separated out for ease of testing."""
-        batch = web.Batch(simulations=simulations, simulation_type="tidy3d_design", **kwargs)
+class MethodSample(Method, ABC):
```
I thought of this when you mentioned the issues with Batch and large number of tasks. Should we maybe provide a `batch_size : pd.PositiveInt = None` parameter? if `None` it does all of the tasks in a single batch. Otherwise it batches them? might be what's needed to handle super large amounts of sampling points

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
     """Select parameters uniformly on a grid.
 
+    Size of the grid is specified by the parameter type,
+    either as the number of unique discrete values (ParameterInt, ParameterAny)
```
variables in docstrings should be in double backslash quotes for RST

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
 
-    def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, Any]:
-        """Defines how the design parameters are sampled on grid."""
+    def get_run_count(self, parameters: list) -> int:
```
docstring

m-bone replied:
Added all these. Does the docstring not inherit from the `@abstractmethod` defined in `Method`?

tylerflex replied:
that's a good question. I dont think it does unfortunately.

m-bone replied:
That's good to know!

tylerflex replied:
the docstrings remain in subclasses as long as you dont overwrite the methods

```
>>> class A:
...     def method(self):
...             """docstring"""
...             pass
...
>>> class B(A):
...     pass
...
>>> B.method.__doc__
'docstring'
>>> class C(A):
...     def method(self):
...             pass
...
>>> C.method.__doc__
(nothing printed)
```

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
+    """A method for handling design searches that optimize the design"""
+
+    # NOTE: We could move this to the Method base class but it's not relevant to MethodGrid
+    rng_seed: pd.PositiveInt = pd.Field(
```
might be fine to just call this `seed`? but not a big deal either way.

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
 
+    acq_func: Literal["ucb", "ei", "poi"] = pd.Field(
+        default="ucb",
+        title="Type of acquisition function",
```
title case for all of these

tylerflex commented on tidy3d/plugins/design/method.py:
```diff
@@ -23,335 +20,780 @@ class Method(Tidy3dBaseModel, ABC):
-        return DEFAULT_MONTE_CARLO_SAMPLER_TYPE(d=d)
+    n_iter: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of bayesian optimization iterations",
```
Title case, and I noticed from time to time `bayesian` not being capitalized. In some outputs, just FYI

tylerflex commented on tidy3d/plugins/design/method.py:
missing period (see other link ends too)

tylerflex commented on tidy3d/plugins/design/method.py:
this method is a bit dense, but maybe that's ok.. I'm having trouble following so maybe either a mild refactor or some inline comments could help

m-bone replied:
Yeah there's become a lot of fiddly code to get Bay opt succesfully through all the different scenarios. I've added some more comments and spacing.

tylerflex commented on tidy3d/plugins/design/method.py:
`isinstance(param, ParameterAny)`, in case we subclass ParameterAny down the line

tylerflex commented on tidy3d/plugins/design/method.py:
will be interesting to see how these look in the docs API reference. Maybe ask Dario to help you compile that locally and just do a read through before we release it.

tylerflex commented on tidy3d/plugins/design/method.py:
docstrings and type annotations, even for these sub functions :D

tylerflex commented on tidy3d/plugins/design/method.py:
I forget, did we want to make `MethodRandom` just subclass from `MethodMonteCarlo` with a warning? or just remove? fine either way.

m-bone replied:
We've removed it now as MonteCarlo is just doing it better.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -7,17 +7,24 @@
 
 class DesignSpace(Tidy3dBaseModel):
-    """Specification of a design problem / combination of several parameters + algorithm.
+    """Manages all exploration of a parameter space within specified parameters using a supplied search method.
```
we're probably going to need a lot more info here since it's the fundamental object for everything. including probably some diagrams and links to notebooks. See Simulation for inspiration

https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/simulation.py#L1557-L1658

m-bone replied:
That's fair. I think I've included a lot of the information in the .run command (including the pre output->post input table we've been talking about) as that's where most people are interacting with the DesignSpace. What do you think should be included specifically here?

tylerflex replied:
I think:
* A few links to our docs examples using the design plugin.
* A code example, minimum viable example of running the DesignSpace (maybe similar to the first few cells of the Design.ipynb where we sample a 2D gaussian function).
* The design plugin flow diagram

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -41,31 +49,51 @@ class DesignSpace(Tidy3dBaseModel):
+    task_name: str = pd.Field(
+        "",
+        title="Task Name",
+        description="Task name assigned to tasks along with a simulation counter in the form of {task_name}_{counter}. \
```
dont need the `\`, we just end the string with `"` (no comma) and continue on next line and it seems to work

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -86,38 +116,281 @@ def get_fn_source(function: Callable) -> str:
+
+        If used as a pre function, the output of ``fn`` must be a float, ``Simulation``, ``Batch``, list, or dict. Supplied ``Batch`` objects are
+        run without modification and are run in series. A list or dict of ``Simulation`` objects is flattened into a single ``Batch`` to enable
+        parallel computation on the cloud. The original structure is then restored for output; all `Simulation`` objects are replaced by ``SimulationData`` objects.
```
careful with the single vs double quotes.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -86,38 +116,281 @@ def get_fn_source(function: Callable) -> str:
+        or a container where the first element is a ``float`` and second element is a ``list`` / ``dict`` e,g. [float {"aux_1": str}].
+        The float is used by the optimizers as the return of the fitness function.
+        The second element is for auxiliary data from the analysis that the user may want to keep.
+        Sampling methods (``MethodGrid`` or ``MethodMonteCarlo``) can have any return type.
```
I dont know if it's just my inability to read properly. but I do think just putting a markdown table might be helpful. For a given `fn_pre` output type, what operation(s) do we do behind the scene, and what type to expect as `fn_post` input.

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -86,38 +116,281 @@ def get_fn_source(function: Callable) -> str:
+            }
+            was_list = True
+
+        def _find_and_map(
```
this is quite complex :D . but I think we discussed it enough. if it's understandable enough to you it's fine. And we can always work on it in the future

m-bone replied:
It certainly is. Do you think it would help moving all the inline func definitions to the top of the file, or to a `design_utils.py`?

tylerflex replied:
hm, I think it might make sense to leave them here actually. since they're easier to reference when reading through. 

some of this complexity is probably unavoidable..

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -128,52 +401,201 @@ def run_batch(
+        else:
+            return round(per_run_estimate * run_count, 3)
+
+    def summarize(self, fn_pre: Callable = None) -> dict[str, Any]:
```
what if I want the summary dict but dont want to print it? `verbose: bool=True` here maybe?

tylerflex reviewed (changes requested):
Thanks @m-bone This is looking great. 80% of my comments were minor proofreading things. I think it's almost done. Only slight concerns right now are:
1. Some of these methods are a bit complex. which is not necessarily bad, but could make them a bit hard to maintain them in long run. If you see any way to simplify them, please take the opportunity now while it's still fresh.
2. I think docs will be very important here for usability, so please flesh out some of the docstrings for the main class with as much info, diagrams, links as you can, including links to the tutorial notebooks.

Other than that and some minor things, looks really good, think this will be a great addition!

tylerflex commented on tidy3d/plugins/design/design.py:
```diff
@@ -41,31 +74,52 @@ class DesignSpace(Tidy3dBaseModel):
+    )
+
+    task_name: str = pd.Field(
+        "",
```
I think we should either make this required or make it something besides ""? What do you think? I'm worried about it being a bit confusing to have task name of "". (was that what it was before?)

m-bone replied:
I've made it like this because task_name is really a prefix the user sets that is combined with the run count from the design process, so it won't be blank in reality. I had previously set it as `None` but this wasn't every checked for and got passed as a string.

m-bone replied:
The docs are also looking good now, I went through and updated them for all the main pages

tylerflex reviewed (commented):
Looks good to me, just not sure about the task_name default. Also just wondering if you checked how the docs look when compiled?

tylerflex reviewed (commented):
looks good! just a few final things:

1. Please add one or more items in the CHANGELOG.md. For example, I think: a sentence under "changed" summarizing the major changes to `design`. 1 or more bullet points under "added" discussing the new optimization methods. Basically this will be our mini advertisement that the tool has these new features
2. Could you squash all of these commits into 1 (or at most, a handful?) reason is to avoid an onslaught of new commits int he public branch. Just one or a few indicating the major changes should be easier. To squash, I usually rebase and "fixup" everything but the top commit, which I "reword" and call something general, eg ("design plugin overhaul and added optimization methods")
3. After this we can merge the notebook PR

Nice work Matt! really excited to see this in production

tylerflex commented on CHANGELOG.md:
```diff
@@ -18,6 +18,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- Added optimization methods to the Design plugin. The plugin has been expanded to include Bayesian optimization, genetic algorithms and particle swarm optimization. Explanations of these methods are available in new and updated notebooks.
+- Added new support functions for the Design plugin: automated batching of `Simulation` objects, and summary functions with `DesignSpace.estimate_cost` and `DesignSpace.summarize`.
 
 ### Changed
```
actually could you add a change item just to let users know of any API changes they might need?
